url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/org/free/bacnet4j/util/IpAddressUtils.java

```diff
@@ -0,0 +1,138 @@
+package org.free.bacnet4j.util;
+
+public class IpAddressUtils{
+
+	public static boolean ipWhiteListCheck(String allowedIp, String remoteIp)throws IpWhiteListException{
+		String remoteIpParts[] = remoteIp.split("\\.");
+		if(remoteIpParts.length != 4)
+			throw new IpWhiteListException((new StringBuilder()).append("Invalid remote IP address: ").append(remoteIp).toString());
+		else
+			return ipWhiteListCheckImpl(allowedIp, remoteIp, remoteIpParts);
+	}
+
+	public static boolean ipWhiteListCheck(String allowedIps[], String remoteIp)
+													throws IpWhiteListException{
+		String remoteIpParts[] = remoteIp.split("\\.");
+		if(remoteIpParts.length != 4)
+			throw new IpWhiteListException((new StringBuilder()).append("Invalid remote IP address: ").append(remoteIp).toString());
+		for(int i = 0; i < allowedIps.length; i++)
+			if(ipWhiteListCheckImpl(allowedIps[i], remoteIp, remoteIpParts))
+				return true;
+		return false;
+	}
+
+	private static boolean ipWhiteListCheckImpl(String allowedIp, String remoteIp, String remoteIpParts[])
+												   throws IpWhiteListException {
+		String allowedIpParts[] = allowedIp.split("\\.");
+		if(allowedIpParts.length != 4)
+			throw new IpWhiteListException((new StringBuilder()).append("Invalid allowed IP address: ").append(allowedIp).toString());
+		else
+			return validateIpPart(allowedIpParts[0], remoteIpParts[0], allowedIp, remoteIp) 
+				&& validateIpPart(allowedIpParts[1], remoteIpParts[1], allowedIp, remoteIp)
+				&& validateIpPart(allowedIpParts[2], remoteIpParts[2], allowedIp, remoteIp)
+				&& validateIpPart(allowedIpParts[3], remoteIpParts[3], allowedIp, remoteIp);
+		}
+
+	private static boolean validateIpPart(String allowed, String remote, String allowedIp, String remoteIp)
+													throws IpWhiteListException{
+		if("*".equals(allowed))
+			return true;
+		int dash = allowed.indexOf('-');
+		int from;
+		int to;
+		int rem;
+		try{
+			if(dash == -1)
+				return Integer.parseInt(allowed) == Integer.parseInt(remote);
+		} catch(NumberFormatException e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)